lib/ClangImporter/ClangModuleDependencyScanner.cpp

```diff
@@ -0,0 +1,341 @@
+    StringRef moduleName, ModuleDependenciesCache &cache) {
+  // Check whether there is already a cached result.
+  if (auto found = cache.findDependencies(
+          moduleName, ModuleDependenciesKind::Clang))
+    return found;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
+
+  // HACK! Replace the module import buffer name with the source file hack.
+  auto importHackFile = clangImpl->getImportHackFile();
+  if (!importHackFile) {
+    // FIXME: Emit a diagnostic here.
+    return None;
+  }
+
+  // Reform the Clang importer options.
+  // FIXME: Just save a reference or copy so we can get this back.
+  ClangImporterOptions importerOpts;
+
+  // Determine the command-line arguments for dependency scanning.
+  auto &ctx = Impl.SwiftContext;
+  std::vector<std::string> commandLineArgs =
+    getClangDepScanningInvocationArguments(
+      ctx, importerOpts, *importHackFile);
+
+  // HACK! Trick out a .m file to use to import the module we name.
+  std::string moduleNameHackDefine =
+      ("-DHACK_MODULE_NAME=" + moduleName).str();
+  commandLineArgs.push_back(moduleNameHackDefine);
+  commandLineArgs.push_back("-fmodules-ignore-macro=HACK_MODULE_NAME");
+
+  std::string workingDir =
+      ctx.SourceMgr.getFileSystem()->getCurrentWorkingDirectory().get();
+  CompileCommand command(workingDir, *importHackFile, commandLineArgs, "-");
+  SingleCommandCompilationDatabase database(command);
+
+  auto clangDependencies = clangImpl->tool.getFullDependencies(
+      database, workingDir, clangImpl->alreadySeen);
+
+  if (!clangDependencies) {
+    // FIXME: Route this to a normal diagnostic.
+    llvm::logAllUnhandledErrors(clangDependencies.takeError(), llvm::errs());
+    return None;
+  }
+
+  // Record module dependencies for each module we found.
+  recordModuleDependencies(cache, *clangDependencies);
+
+  return cache.findDependencies(moduleName, ModuleDependenciesKind::Clang);
+}
+
+bool ClangImporter::addBridgingHeaderDependencies(
+    StringRef moduleName,
+    ModuleDependenciesCache &cache) {
+  auto targetModule = *cache.findDependencies(
+      moduleName, ModuleDependenciesKind::Swift);
+
+  // If we've already recorded bridging header dependencies, we're done.
+  auto swiftDeps = targetModule.getAsSwiftModule();
+  if (!swiftDeps->bridgingSourceFiles.empty() ||
+      !swiftDeps->bridgingModuleDependencies.empty())
+    return false;
+
+  // Retrieve or create the shared state.
+  auto clangImpl = getOrCreateClangImpl(cache);
+
+  // Reform the Clang importer options.
+  // FIXME: Just save a reference or copy so we can get this back.
+  ClangImporterOptions importerOpts;
+
+  // Retrieve the bridging header.
+  std::string bridgingHeader = *targetModule.getBridgingHeader();
+
+  // Determine the command-line arguments for dependency scanning.
+  auto &ctx = Impl.SwiftContext;
+  std::vector<std::string> commandLineArgs =
```

Comment:
Sorry if all these  `consts` are getting tedious. As you know, the idea is that if there are more than a few lines after the assignment, if something is `const` the first assignment is enough to know where it's coming from without having to inspect all uses. (Esp. in C++ where one can pass a reference, sigh.) Forgive me for stating the obvious.